File: src/main/java/gargoyle/ct/ui/util/render/FunctionalListCellRenderer.java
package gargoyle.ct.ui.util.render;

import javax.swing.*;
import java.awt.*;
import java.util.function.Function;

public class FunctionalListCellRenderer<E> implements ListCellRenderer<E> {
    private final Function<E, String> messageProvider;
    private final ListCellRenderer originalRenderer;

    public FunctionalListCellRenderer(final ListCellRenderer originalRenderer,
                                      final Function<E, String> messageProvider) {
        this.originalRenderer = originalRenderer;
        this.messageProvider = messageProvider;
    }

    public Component getListCellRendererComponent(
            JList<? extends E> list,
            E value,
            int index,
            boolean isSelected,
            boolean cellHasFocus) {
        String message = messageProvider.apply(value);
        //noinspection unchecked
        return originalRenderer.getListCellRendererComponent(list, message, index, isSelected, cellHasFocus);
    }
}

File: src/main/kotlin/gargoyle/ct/ui/util/render/FunctionalListCellRenderer.kt
package gargoyle.ct.ui.util.render

import java.awt.Component
import java.util.function.Function
import javax.swing.JList
import javax.swing.ListCellRenderer

open class FunctionalListCellRenderer<E : Any>(
    private val originalRenderer: ListCellRenderer<E>,
    private val messageProvider: Function<E, String>
) : ListCellRenderer<E> {
    @Suppress("UNCHECKED_CAST")
    override fun getListCellRendererComponent(
        list: JList<out E>,
        value: E,
        index: Int,
        isSelected: Boolean,
        cellHasFocus: Boolean
    ): Component {
        val message = messageProvider.apply(value) as E
        return originalRenderer.getListCellRendererComponent(list, message, index, isSelected, cellHasFocus)
    }
}
